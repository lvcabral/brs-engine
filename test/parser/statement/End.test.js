const brs = require("../../../bin/brs.node");
const { Lexeme } = brs.lexer;

const { token, identifier, EOF } = require("../ParserTests");

describe("`end` statement", () => {
    beforeEach(() => {
        parser = new brs.parser.Parser();
    });

    it("does not produce errors", () => {
        let { tokens } = brs.lexer.Lexer.scan(`
                sub Main()
                    end
                end sub
            `);
        let { statements, errors } = parser.parse(tokens);
        expect(errors.length).toEqual(0);
        expect(statements).toMatchSnapshot();
    });

    it("is valid as a statement", () => {
        let { statements, errors } = brs.parser.Parser.parse([token(Lexeme.End, "end"), EOF]);
        expect(errors[0]).toBeUndefined();
        expect(statements).toMatchSnapshot();
    });

    it("can be used as a property name on objects", () => {
        let { tokens } = brs.lexer.Lexer.scan(`
                sub Main()
                    person = {
                        end: true
                    }
                end sub
            `);
        let { statements, errors } = parser.parse(tokens);
        expect(errors.length).toEqual(0);
        expect(statements).toMatchSnapshot();
    });

    it("is not allowed as a standalone variable", () => {
        //this test depends on token locations, so use the lexer to generate those locations.
        let { tokens } = brs.lexer.Lexer.scan(`
                sub Main()
                    end = true
                end sub
            `);
        let { statements, errors } = parser.parse(tokens);
        expect(errors.length).toEqual(1);
        //specifically check for the error location, because the identifier location was wrong in the past
        expect(errors[0].location).toEqual({
            file: "",
            start: { line: 3, column: 24 },
            end: { line: 3, column: 25 },
        });
        expect(statements).toMatchSnapshot();
    });
});
