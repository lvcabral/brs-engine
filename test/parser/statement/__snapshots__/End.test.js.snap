// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`end\` statement can be used as a property name on objects 1`] = `
[
  Function {
    "func": Function {
      "body": Block {
        "location": {
          "end": {
            "column": 16,
            "line": 6,
          },
          "file": "",
          "start": {
            "column": 26,
            "line": 2,
          },
        },
        "statements": [
          Assignment {
            "name": {
              "isReserved": false,
              "kind": "Identifier",
              "literal": undefined,
              "location": {
                "end": {
                  "column": 26,
                  "line": 3,
                },
                "file": "",
                "start": {
                  "column": 20,
                  "line": 3,
                },
              },
              "text": "person",
            },
            "tokens": {
              "equals": {
                "isReserved": false,
                "kind": "Equal",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 28,
                    "line": 3,
                  },
                  "file": "",
                  "start": {
                    "column": 27,
                    "line": 3,
                  },
                },
                "text": "=",
              },
            },
            "value": AALiteral {
              "close": {
                "isReserved": false,
                "kind": "RightBrace",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 21,
                    "line": 5,
                  },
                  "file": "",
                  "start": {
                    "column": 20,
                    "line": 5,
                  },
                },
                "text": "}",
              },
              "elements": [
                {
                  "name": BrsString {
                    "inArray": false,
                    "kind": 3,
                    "value": "end",
                  },
                  "value": Literal {
                    "_location": {
                      "end": {
                        "column": 33,
                        "line": 4,
                      },
                      "file": "",
                      "start": {
                        "column": 29,
                        "line": 4,
                      },
                    },
                    "value": BrsBoolean {
                      "inArray": false,
                      "kind": 2,
                      "value": true,
                    },
                  },
                },
              ],
              "open": {
                "isReserved": false,
                "kind": "LeftBrace",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 30,
                    "line": 3,
                  },
                  "file": "",
                  "start": {
                    "column": 29,
                    "line": 3,
                  },
                },
                "text": "{",
              },
            },
          },
        ],
      },
      "end": {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 23,
            "line": 6,
          },
          "file": "",
          "start": {
            "column": 16,
            "line": 6,
          },
        },
        "text": "end sub",
      },
      "keyword": {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 19,
            "line": 2,
          },
          "file": "",
          "start": {
            "column": 16,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": [],
      "returns": 11,
    },
    "name": {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": {
        "end": {
          "column": 24,
          "line": 2,
        },
        "file": "",
        "start": {
          "column": 20,
          "line": 2,
        },
      },
      "text": "Main",
    },
  },
]
`;

exports[`\`end\` statement does not produce errors 1`] = `
[
  Function {
    "func": Function {
      "body": Block {
        "location": {
          "end": {
            "column": 16,
            "line": 4,
          },
          "file": "",
          "start": {
            "column": 26,
            "line": 2,
          },
        },
        "statements": [
          End {
            "tokens": {
              "end": {
                "isReserved": false,
                "kind": "End",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 23,
                    "line": 3,
                  },
                  "file": "",
                  "start": {
                    "column": 20,
                    "line": 3,
                  },
                },
                "text": "end",
              },
            },
          },
        ],
      },
      "end": {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 23,
            "line": 4,
          },
          "file": "",
          "start": {
            "column": 16,
            "line": 4,
          },
        },
        "text": "end sub",
      },
      "keyword": {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 19,
            "line": 2,
          },
          "file": "",
          "start": {
            "column": 16,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": [],
      "returns": 11,
    },
    "name": {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": {
        "end": {
          "column": 24,
          "line": 2,
        },
        "file": "",
        "start": {
          "column": 20,
          "line": 2,
        },
      },
      "text": "Main",
    },
  },
]
`;

exports[`\`end\` statement is not allowed as a standalone variable 1`] = `
[
  Function {
    "func": Function {
      "body": Block {
        "location": {
          "end": {
            "column": 16,
            "line": 4,
          },
          "file": "",
          "start": {
            "column": 26,
            "line": 2,
          },
        },
        "statements": [
          End {
            "tokens": {
              "end": {
                "isReserved": false,
                "kind": "End",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 23,
                    "line": 3,
                  },
                  "file": "",
                  "start": {
                    "column": 20,
                    "line": 3,
                  },
                },
                "text": "end",
              },
            },
          },
        ],
      },
      "end": {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 23,
            "line": 4,
          },
          "file": "",
          "start": {
            "column": 16,
            "line": 4,
          },
        },
        "text": "end sub",
      },
      "keyword": {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 19,
            "line": 2,
          },
          "file": "",
          "start": {
            "column": 16,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": [],
      "returns": 11,
    },
    "name": {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": {
        "end": {
          "column": 24,
          "line": 2,
        },
        "file": "",
        "start": {
          "column": 20,
          "line": 2,
        },
      },
      "text": "Main",
    },
  },
]
`;

exports[`\`end\` statement is valid as a statement 1`] = `
[
  End {
    "tokens": {
      "end": {
        "isReserved": false,
        "kind": "End",
        "literal": undefined,
        "location": {
          "end": {
            "column": -9,
            "line": -9,
          },
          "start": {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end",
      },
    },
  },
]
`;
